L73-java-binderchannel/wireformat.md

```diff
@@ -0,0 +1,173 @@
+    *   From ClientTransport to server binder.
+*   Ping Response:
+    *   Transaction code: BinderTransport.PING_RESPONSE
+    *   From SeverTransport to client binder, in response to a ping.
+
+#### Transaction Data
+
+```
+version = int; (* Currently always 1 *)
+num bytes = int;
+ping id = int;
+
+setup transport transaction = version, binder;
+shutdown transport transaction = nothing;
+acknowledge bytes transaction = num bytes;
+ping transaction = ping id;
+ping response transaction = ping id;
+```
+
+### Stream Transactions
+
+Aside from the reserved 1000 transaction codes, all others are used for streams.
+When the ClientTransport creates a new Stream, it assigns an ID to use for all
+transactions relating to that stream (in both directions).
+
+IDs are allocated sequentially from FIRST_CALL_ID to LAST_CALL_ID, wrapping
+around at the end. Since there are roughly 2^24 available IDs, there’s no chance
+of a collision except in pathological cases. If we _do_ detect a collision
+however, we immediately shut down the transport with an INTERNAL error code.
+
+Each normally-completed stream transmits (in both directions) a prefix, 0 or
+more blocks of message data, and a suffix. The contents of prefix and suffix
+data are direction dependenet, and all three may can be sent in a single
+transaction.
+
+#### Transaction Data
+
+```
+flags = int; (* A bitset of flags, and an optional status code. *)
+sequence no = int; (* A transaction sequence no for the RPC *)
+method ref = string; (* The full method name *)
+status = [string] (* The status code is in the flags. The description is only
+                     included if FLAG_STATUS_DESCRIPTION is set.  *);
+count = int;
+bytes data = count, [byte array]; (* byte array included iff count > 0 *)
+metadata = count, { metadata key, metadata val }; (* repeated count times *)
+metadata sentinel = int (* -1 = Parcelable, -2 = bad metadata. *)
+metadata key = bytes data;
+metadata value = (count | metadata sentinel), [parcelable], [byte array];
+
+client prefix = method ref, metadata
+client suffix = (* currently empty *)
+
+server prefix = metadata (* headers *)
+server suffix = status, metadata (* trailers *)
+
+message data = parcelable | bytes data (* parcelable iff
+                                          FLAG_MESSAGE_DATA_IS_PARCELABLE
+                                          set, bytes otherwise *)
+client rpc data = [client prefix] (* if FLAG_PREFIX is set *)
+                  [message data]  (* if FLAG_MESSAGE_DATA is set *)
+                  [client suffix] (* if FLAG_SUFFIX is set *)
+server rpc data = [server prefix] (* if FLAG_PREFIX is set *)
+                  [message data]  (* if FLAG_MESSAGE_DATA is set *)
+                  [server suffix] (* if FLAG_SUFFIX is set *)
+out of band close data = status;
+rpc data = client rpc data | server rpc data | out of band close data;
+
+rpc transaction = flags, sequence no, rpc data;
+```
+
+#### Flag fields
```

Comment:
We should define what happens if an unknown field is present.

Reply:
ignore unrecognized.